peekingduck/nodes/output/media_writer.py

```diff
@@ -126,3 +169,12 @@ def _append_datetime_filename(self, filename: str) -> str:
         file_path_with_timestamp = self.output_dir / filename_with_timestamp
 
         return str(file_path_with_timestamp)
+
+    def _append_frame_filename(self, filename: Optional[str]) -> str:
+        """append frame to filename before extension Format: filename_frame.extension"""
+        assert filename is not None, "filename cannot be None"
+        filename_with_frame = f"_{self._frame:05d}.".join(filename.split(".")[-2:])
+        file_path_with_frame = self.output_dir / filename_with_frame
+        self._frame += 1
+
+        return str(file_path_with_frame)
```

Comment:
When testing your code, I got the following error:
```
* input_source=videos/pkd_hero_video.mp4
2022-12-22 20:17:12 peekingduck.nodes.input.visual  INFO:  Input size: 854 by 480 
[ERROR:0@9.124] global /Users/xperience/actions-runner/_work/opencv-python/opencv-python/opencv/modules/videoio/src/cap.cpp (597) open VIDEOIO(CV_IMAGES): raised OpenCV exception:

OpenCV(4.6.0) /Users/xperience/actions-runner/_work/opencv-python/opencv-python/opencv/modules/videoio/src/cap_images.cpp:253: error: (-5:Bad argument) CAP_IMAGES: can't find starting number (in the name of file): my_output/my_video in function 'icvExtractPattern'
```
First line is specially printed out by me for debugging.
Did you encounter this during testing?

Reply:
Have no issue with video in local test.
`tests/nodes/output/test_media_writer.py` updated to test this new feature.

Reply:
I am still getting the above error with the latest code.
What is your M1 OpenCV version?

The pipeline I use is:
```
❯ cat output_filename.yml
nodes:
- input.visual:
    source: videos
- output.screen:
- output.media_writer:
    output_dir: my_output
    output_filename: my_video
```

I use a folder `videos` of 4 test videos, that can be found here:
https://drive.google.com/drive/folders/1Ge8hGdVTYgTdOl-m8QuSoiteFuupQ6b5?usp=share_link

Reply:
I have change the pipeline to below:

```yaml
nodes:
- input.visual:
    source: "data/"
- output.screen:
- output.media_writer:
    output_dir: "data/my_output"
    output_filename: my_video.mp4
```

I have to add the `.mp4` extension in attr `output_filename` as it is required by OpenCV. Using M1 env `opencv                    4.5.5            py38h150bfb4_7    conda-forge`, the video are output correctly.

Reply:
Thanks for testing. Turns out the missing `.mp4` extension in the output filename is causing the OpenCV error.
I would suggest doing some defensive programming: check that the output filename contains a valid extension, and if none exists, use `.mp4` as default.

Reply:
function `_validate_file_extension()` implemented to check and return filename with valid extension.

Reply:
refactor to `_get_outputfile_extension(input_filename: str, output_filename: str) -> str:`

For empty output filename extension, this function will follow input filename extension `output_filename + Path(input_filename).suffix`. This is to to prevent image->video scenario.